merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts

```diff
@@ -8,20 +9,24 @@ export class Constraint {
     this.__astNode = astNode;
   }
 
-  public static ForbiddenActivityOverlap(activityType1: string, activityType2: string): Constraint {
+  public static ForbiddenActivityOverlap(activityType1: Gen.ActivityTypeName, activityType2: Gen.ActivityTypeName): Constraint {
     return new Constraint({
       kind: AST.NodeKind.ForbiddenActivityOverlap,
       activityType1,
       activityType2
     })
   }
 
-  public static ForEachActivity(activityType: string, alias: string, constraint: Constraint): Constraint {
+  public static ForEachActivity<A extends Gen.ActivityTypeName>(activityType: A, expression: (instance: Gen.ActivityInstance<A>) => Constraint): Constraint {
+    let expressionString = expression.toString();
+    let startIndex = expressionString.indexOf("(") + 1;
+    let endIndex = expressionString.indexOf(")");
+    let alias = expressionString.substring(startIndex, endIndex);
     return new Constraint({
       kind: AST.NodeKind.ForEachActivity,
       activityType,
       alias,
-      expression: constraint.__astNode
+      expression: expression(new Gen.ActivityInstance(activityType, alias)).__astNode
```

Comment:
What does a call of this look like?

PS There are vulnerabilities in this expressionString implementation. Example: its perfectly valid to wrap a callback function in parens, so for example

`[1,2,3].map(((((num) => num + 1))))` would typecheck fine.

Or you can omit the parens for an arrow function

`[1,2,3].map(num => num + 1)` would typecheck fine.

Reply:
🤦 Didn't see the example in the descriptions. I don't think we need the alias at all in this case - or rather we can have an auto-generated alias with this formulation.

Reply:
I've made that change as an extra commit that I'll either squash or revert later.